Use CentredInput in SearchForm

mstriemer opened the pull request:
This was how we wanted the input to look on Marketplace so I'm guessing we want to do something similar here.

I put `<CentredInput />` in the `src/ui/components` folder, I'm not sure if we want to name that something else. I think I like the idea of having an app that's just reusable components though. Perhaps the magnifying glass could get moved into a wrapper component called `<CentredSearchInput />` or something.

We could probably make a little app in `src/ui` to be a styleguide and show off and document the components as well.

![centred-search-form mov](https://cloud.githubusercontent.com/assets/211578/19182972/7d08bb06-8c76-11e6-92e6-041fa9a9d8ad.gif)

@pwalm thoughts?

tofumatt commented:
It looks like it's ever-so-slightly to the left and not centred, maybe it's just the magnifying glass making it look that way and it needs some left-padding?

Otherwise I think this is way better than what I had.

tofumatt commented on src/ui/components/CentredInput/style.scss:
```diff
@@ -0,0 +1,32 @@
+  width: 100%;
+}
+
+.CentredInput-input::placeholder {
```
Since this is `.scss` now (I know it wasn't in the original repo), could we nest these? I find it easier to scan 😄

tofumatt commented on src/ui/components/CentredInput/style.scss:
```diff
@@ -0,0 +1,32 @@
+  background: transparent;
+  border: none;
+  box-sizing: border-box;
+  left: 0;
```
Does this need adjusting for RTL?

tofumatt commented on src/amo/components/SearchForm.scss:
```diff
@@ -1,48 +1,58 @@
+  // text is entered; we do this because centring the text takes precedence
+  // over showing the magnifying glass, and we can't do that (easily) with
+  // search text that can expand.
+  background: url('~amo/img/icons/magnifying-glass.svg') 0 50% no-repeat;
```
Will the `::before` put this to the right on RTL? I think so, I just forget.

tofumatt commented on src/amo/components/SearchForm.scss:
```diff
@@ -1,48 +1,58 @@
+  text-overflow: ellipsis;
+}
+
+.SearchForm-query .CentredInput-label:before {
```
I think this should be `::before`, since we aren't targetting any browsers that don't support the new notation.

Also, would be nice to nest.

tofumatt commented on src/amo/components/SearchForm.scss:
```diff
@@ -1,48 +1,58 @@
+.SearchForm-query.CentredInput,
+.SearchForm-query .CentredInput-label,
+.SearchForm-query .CentredInput-input {
+  @include font-regular();
```
Oh, nice. 👍

tofumatt commented on src/amo/components/SearchForm.scss:
```diff
@@ -1,48 +1,58 @@
+  background: url('~amo/img/icons/magnifying-glass.svg') 0 50% no-repeat;
+  content: '';
+  display: inline-block;
+  left: -$search-query-padding;
```
Does this need an RTL tweak as well?

tofumatt commented on src/ui/components/CentredInput/index.js:
```diff
@@ -0,0 +1,58 @@
+
+class CentredInput extends React.Component {
+  hidePlaceholder() {
+    this.label.style.display = 'none';
```
Can we do this with a class instead? Seems cleaner to me.

mstriemer commented:
> It looks like it's ever-so-slightly to the left and not centred, maybe it's just the magnifying glass making it look that way and it needs some left-padding?

I centred the text and put the icon beside it rather than centring the two together. I was going to ask @pwalm for his thoughts and mocked them both out. I now think both centred looks better though, but I'll post the pictures anyway.

<img width="411" alt="screenshot 2016-10-11 16 20 45" src="https://cloud.githubusercontent.com/assets/211578/19289412/4bafd794-8fd0-11e6-869a-2eae41edd34d.png">
> Text centred

<img width="411" alt="screenshot 2016-10-11 16 27 06" src="https://cloud.githubusercontent.com/assets/211578/19289411/4bab0b88-8fd0-11e6-8242-982a2ef66e91.png">
> Both centred

tofumatt commented:
Both centred looks better, aside from the fact that it's not vertically centred (it's a bit too high).

mstriemer commented:
So I rebased this and updated it so that the icon and text are centred together. Unfortunately that means that the entered text no longer lines up with the placeholder text.

![centred-input-sort-of mov](https://cloud.githubusercontent.com/assets/211578/19582564/2e40a940-96fb-11e6-81b5-ec3cc8b4b735.gif)

I'm not sure what the right thing to do here is. I'm going to leave this for tonight but I might see about keeping the magnifying glass around tomorrow, that should keep everything lining up but is tricky to accomplish.

muffinresearch commented:
> I'm not sure what the right thing to do here is. I'm going to leave this for tonight but I might see about keeping the magnifying glass around tomorrow, that should keep everything lining up but is tricky to accomplish.

Does this notice as much if your search terms don't match the placeholder?

pwalm commented:
Search flow below! (and "both centred" looks better).

![search flow - xxhdi](https://cloud.githubusercontent.com/assets/6797635/19605547/cfefeb42-9788-11e6-9462-5ef0cd89e1e0.png)

tofumatt commented:
I think that looks like actually, though if you want to add the search icon
feel free :-)
